runtime/src/threads/pthreads/threads-pthreads.c

```diff
@@ -82,6 +98,141 @@ static void*           initial_pthread_func(void*);
 static void*           pthread_func(void*);
 
 
+#if defined(CHPL_USING_CSTDLIB_MALLOC) && defined(__APPLE__)
+#define APPLE__CSTDLIB__ALLOC
+#endif
+
+/* We have to see if the thread stacks should be allocated in
+ * the chapel heap and if we need to add a guard page for catch
+ * a stack overflow
+ */
+
+static void init_heap_stack(void){
+
+  // We don't want a guard page, but we want the stack in the heap
+  if (CHPL_STACK_CHECKS == 0) {
+    alloc_stack_in_heap = true;
+    use_guard_page      = false;
+    return;
+  }
+
+  /* Special case: we want a guard page, but we can't add it. So
+   * we use a pthread stack that provides a guard page.
+   */
+#ifdef APPLE__CSTDLIB__ALLOC
+  alloc_stack_in_heap = false;
+  use_guard_page      = false;
+  return;
+#else
+  else if (chpl_getHeapPageSize() != chpl_getSysPageSize()) {
+    alloc_stack_in_heap = false;
+    use_guard_page      = false;
+    return ;
+  } 
+#endif
+
+  // We want a guard page and we can add it
+  alloc_stack_in_heap = true;
+  use_guard_page      = true;
+}
+
+static void* alloc_pthread_stack(size_t stack_size){
+  thread_stack_list_p          tslp;
+  size_t page_size;
+  size_t mem_size;
+  void* stack;
+  void* mem_buffer;
+  int rc;
+
+  page_size = chpl_getSysPageSize();
+  mem_size = (use_guard_page ? stack_size + page_size : stack_size);
+
+  mem_buffer = chpl_memalign(page_size, mem_size);
```

Comment:
// uses memalign to align to page_size to simplify logic for the guard page case
// but thread stacks could be allocated without such alignment